### ISSUE TITLE
An abstraction that interfaces with any API and uses an LLM to come up with API calls to be executed.

### ISSUE THREAD

**Issue by andreibondarev** [patterns-ai-core/langchainrb#81]
## Background
A popular use-case is constructing a prompt for the LLM that includes the API specs, and user's question at hand to construct an API call that gets executed on LLM's behalf. The API call's result then gets fed back to the LLM to synthesize an answer to the user's question.

### Objectives
Create an entity (should it be a Tool? an Agent? or another abstraction?) that:
- Can execute API calls
- Can construct a prompt for the LLM that:
  - Includes API specs, or only relevant endpoints? Subject to context window limits.
  - Includes the instructions for the LLM
  - Includes the user's question.

**Comment by bborn** [comment]
I've spent quite a bit of time on this (in Python). Here's what I came up with for AgentHQ:

https://gist.github.com/bborn/59ec195828825445e97f8eb37e8fbc2b

Basically:

- Get the OpenAPI spec (yaml/json)
- Convert the spec to a single page markdown documentation (using `widdershins`)
- Ingest the markdown doc and create a vector index from it (LlamaIndex)
- Query the vector index, asking for the correct python code snippet required to answer the user's query/complete their request.
- If the request fails, save the error message, stuff it back into the prompt (e.g. "here's what you tried before that didn't work, and here's the error") and then try again (with some max_retry limit)
